connection.c

```diff
@@ -127,13 +128,63 @@ void ConnAsyncTerminate(Connection *conn)
     conn->flags |= CONN_TERMINATING;
 }
 
+static void handleAuth(redisAsyncContext *c, void *r, void *privdata)
+{
+    Connection *conn = privdata;
+    JoinLinkState *state = ConnGetPrivateData(conn);
+
+    redisReply *reply = r;
+    if (!reply) {
+        LOG_ERROR("Redis connection authentication failed: connection died");
+        ConnMarkDisconnected(conn);
+    } else if (reply->type == REDIS_REPLY_ERROR) {
+        LOG_ERROR("Redis connection authentication failed: %s", reply->str);
+        state->failed = true;
```

Comment:
Once we set failed, will it stay failed forever? Not just the usage here but in general, about marking a connection failed, should it work again once we fix the issue? For this password issue, after we change the password locally or remotely, should it work? Otherwise, admin must restart the node? No?

Another issue is logging, say I misconfigured the node, I'll take a look at the node an hour later, will I see a big log file with same lines over an over? Probably, with `unrecoverable error, check logs` lines which is printed in `joinLinkIdleCallback()`. (I don't know, I didn't try).

Probably, these are not important issues, maybe I'm wrong. Just saying, maybe we should take a look at Redis behavior and mimic it for these.

Reply:
if it fails auth, we're in a bad state.  but yes, need to think about how it can recover from that.

Reply:
so dug into this.  the failed boolean is only used by join/link.  it is not used by normal reconnection (for already linked nodes).  so I think its fine to leave them there, if the connection fails, join/link will fail and return to the user and this connection object will be killed.  If already connected, it will be ignored and will try to reconnect.

with that said, want to think about it a bit more to see if we can improve connection handling in general.

Reply:
as noted, failed is gone, still not 100% believing that I handle the connection correctly, but think its correct.